packages/iov-ethereum/src/abi.ts

```diff
@@ -0,0 +1,89 @@
+import BN = require("bn.js");
+
+import { Address } from "@iov/bcp";
+import { Encoding } from "@iov/encoding";
+import { isValidAddress } from "./address";
+
+export interface HeadTail {
+  /** An array of start positions within the original data */
+  readonly head: ReadonlyArray<number>;
+  /** Arguments split by positions as defined by head */
+  readonly tail: ReadonlyArray<Uint8Array>;
+}
+
+export class Abi {
```

Comment:
You really want to start re-implementing all the abi code?

Why not just copy in relevant files from ether.js?

Reply:
The general ABI encoder/decoder is a complex and spread across mutiple files and packages. It contains a functionality that we don't need: encoding/decoding based on the ABI file. With ERC20 and Atomic swap we have less than 10 calls to covers and can easily compose the call and decompose the result in those places. I'm very happy that non of the code in IOV-Core requires the abi JSON of a contrect (it just exists in the deployment part). Thus I don't think we need a fully featured ABI coder.

Reply:
This is okay for erc20 (all tokens) and atomic swap support (only our contract).

There are requests for general smart contract calls on iov-core (various sales/marketing contracts that consider using iov-core to connect to their private blockchains). They would need some sort of `etherum/raw` transaction type that accepted a `json abi` and `any[]` data.

We don't need it for these stories, but will in the future, and I assumed easier to add that as a basis. However, if you can implement the needed calls easily hard-coded, just go for it, and general smart-contract support will be done in some months when MvP is finished and the requirement officially comes from business.